rfcs/regex_feature.md

```diff
@@ -0,0 +1,89 @@
+# Regex feature
+
+A feature based on regex pattern matching that MindMeld's NER models can use to detect custom entities. 
+
+- Developers expect NER models to correctly infer that the query `AVET128-321` is a `pid` entity 
+without providing surrounding sentence context to the classifier. This is more relevant in services that
+just use MindMeld for NLP while doing their own dialog management. 
+- Developers also want granular distinctions between entities that occur in the same context, 
+for example, `look up {SKU120|sku}` and `look up {123-ABC-12|pid}`. Currently, we ask developers to defer
+these types of distinction to the business logic in the dialogue manager.
+- DialogFlow, NLTK and other similar platforms have the regex matching feature. Moreover, we already have an 
+orthographic matching feature in MindMeld. This feature would augment that.
+
+
+## Usage
+
+The developer makes a new file called `regex_matches.txt` for the entity type in the entities folder that needs 
+these regex matches like this following:
+
+```text
+entities
+├── account_type
+│   ├── gazetteer.txt
+│   ├── mapping.json
+│   └── regex_matches.txt
+├── autopay_status
+│   └── mapping.json
+└── credit_amount
+    └── mapping.json
+```
+
+Inside the `regex_matches.txt` file, the developer lists regex patterns compliant with python syntax that need 
+to be matched for this custom entity. 
+
+```text
+\w{4}[0-9]{3}\-[0-9]{3}
+[0-9]{3}\-[0-9]{3}
+```
+
+In the above syntax, the `account_type` entity would match words like `ABCD-123-281` (first pattern) 
+and `123-456` (second pattern).
+
+Finally, the developer would have the ability to turn on/off the regex feature through the model config in 
+config.py using the key,value `regex-seq: True`. The feature will be off by default.
```

Comment:
We should consider providing a warning message if "regex-seq" is off but "regex_matches.txt" files are found. A developer may create regex rules but forget to set "regex-seq" to True since it is off by default.

Reply:
I think this makes sense. One thing to note is that we currently dont warn devs about other entity features like `in-gaz` when the gaz file is present, for example. Ideally, we should just do this intelligently, ie if the `regex_matches.txt`, use it as a feature. But the issue with this is it becomes confusing to developers who don't want to use it even if they have the `regex_matches.txt` file.